### ISSUE TITLE
render/gles2: add support for GL_EXT_window_rectangles

### ISSUE THREAD

**Issue by emersion** [swaywm/wlroots#2644]
Right now we allow compositors to set a scissor region for damage tracking purposes. For each rectangle they need to repaint, compositors repeatedly render the same texture.

radeonsi and nouveau both support [GL_EXT_window_rectangles](https://www.khronos.org/registry/OpenGL/extensions/EXT/EXT_window_rectangles.txt), which basically allows one to define multiple scissor regions. We wouldn't need to repeatedly render textures anymore.

Note, Vulkan has something similar with [VK_EXT_discard_rectangles](https://www.khronos.org/registry/vulkan/specs/1.0-extensions/html/vkspec.html#VK_EXT_discard_rectangles).

**Comment by imirkin** [comment]
A common pattern in sway is to basically do

```
for each rect:
  scissor(rect)
  render()
```

I'm thinking that we can create some variants of `wlr_render_subtexture_with_matrix` & co which take the list of rects directly, and iterate over them in the best way, be it scissors, or rectangles, in manageable groups (e.g. if you have 100 rects, and a max of 4 rectangles, it'll still be 25 draws).

**Comment by imirkin** [comment]
Another alternative is to make it a bit more configurable with macros, like

```
WLR_WITH_RECTS(nrects, rects) {
  wlr_render_subtexture_with_matrix(...);
}
```

Which internally will split this up into groups of draws. This allows the drawing do whatever and not have to support rects directly. Not sure if this fits with the style of wlr though. Opinions welcome.

**Comment by imirkin** [comment]
Another thought is that if you really do have a ton of these rects, drawing them onto a stencil texture and then enabling a stencil test with the final draw is probably going to be faster. I guess I'm not too clear on the circumstances in which there would be multiple rects.

**Comment by emersion** [comment]
> I'm thinking that we can create some variants of wlr_render_subtexture_with_matrix & co which take the list of rects directly, and iterate over them in the best way, be it scissors, or rectangles, in manageable groups (e.g. if you have 100 rects, and a max of 4 rectangles, it'll still be 25 draws).

Yeah, that'd be possible. A less intrusive change would be to make `wlr_renderer_scissor` take a list of rectangles instead of a single rect. We'd still need to handle iterating over the rectangles inside the renderer.

>WLR_WITH_RECTS(nrects, rects)

In general we try to reduce magic macros to a minimum.

> I guess I'm not too clear on the circumstances in which there would be multiple rects.

There should be a small number of rectangles.

These rectangles are damaged regions, ie. regions of our buffer that need to be repainted because they changed. Clients tell us which parts of their buffers changed, and we try to only blit these into our own buffer. Highly granular damage with a lot of rectangles doesn't improve performance a lot, so [we cap the max number of rectangles to 20](https://github.com/swaywm/wlroots/blob/7bc8dbb991ecebd7c5117776a043f886593cbd97/types/wlr_output_damage.c#L102) by default.

**Comment by imirkin** [comment]
> A less intrusive change would be to make `wlr_renderer_scissor` take a list of rectangles instead of a single rect. We'd still need to handle iterating over the rectangles inside the renderer.

OK, so let's say my current code is:

```
for (int i = 0; i < nrects; ++i) {
	wlr_renderer_scissor(renderer, &rects[i]);
	wlr_renderer_clear(renderer, clear_color);
}
```

Is your proposal to change this into

```
wlr_renderer_scissors(renderer, rects);
wlr_renderer_clear(renderer, clear_color);
```

And teach wlr_renderer_clear() to "respect" the list of rects as best it can given their quantity and GL implementation limitations? IMHO that's a bit intense -- first of all, this will require modifying all the rendering functions, but secondly if the loop has multiple render calls, then you're going to be iterating per-rect per-draw-function rather than just per-rect.

That's why I thought of the macro. However if you're allergic to macros, one could also invent e.g. a `wlr_renderer_scissor_batch` which takes an array of rects and produces a `wlr_scissors` object, which would contain a list of rects that an implementation can apply in a single go. And then the new code would look like

```
batches = wlr_renderer_scissor_batch(renderer, rects, &nbatches);
for (int i = 0; i < nbatches; i++) {
	wlr_renderer_scissors(&batches[i]);
	wlr_renderer_clear(renderer, clear_color);
}
```

A sufficiently clever renderer might even internally create the stencil surface if it feels like the number of rects is too high, although I'm not sure that it's worth optimizing things to such a point (and whether they'd actually be any more optimal).

**Comment by emersion** [comment]
As a side note, the Pixman renderer can take an arbitrary clip region.
